File: 251121_last/backend/src/services/weather_service.py
"""
Weather Service - 날씨 비즈니스 로직
"""

import requests
from datetime import datetime, timedelta
import math
from config import Config


class WeatherService:
    # 날씨 비즈니스 로직 계층

    def __init__(self):
        self.api_key = Config.WEATHER_API_KEY
        self.api_url = Config.WEATHER_API_URL

    def get_weather(self, lat: float = 37.5665, lon: float = 126.9780) -> dict:
        # 날씨 정보 조회
        try:
            nx, ny = self._convert_to_grid(lat, lon)
            base_date, base_time = self._get_base_time()

            params = {
                "serviceKey": self.api_key,
                "pageNo": "1",
                "numOfRows": "100",
                "dataType": "JSON",
                "base_date": base_date,
                "base_time": base_time,
                "nx": nx,
                "ny": ny,
            }

            response = requests.get(self.api_url, params=params, timeout=10)
            response.raise_for_status()

            data = response.json()

            if data["response"]["header"]["resultCode"] != "00":
                raise Exception(f"API 에러: {data['response']['header']['resultMsg']}")

            return self._parse_weather_data(data, lat, lon, nx, ny)

        except Exception as e:
            raise Exception(f"날씨 조회 실패: {str(e)}")

    def _convert_to_grid(self, lat: float, lon: float) -> tuple:
        # 위경도를 격자 좌표로 변환
        RE = 6371.00877
        GRID = 5.0
        SLAT1 = 30.0
        SLAT2 = 60.0
        OLON = 126.0
        OLAT = 38.0
        XO = 43
        YO = 136

        DEGRAD = math.pi / 180.0

        re = RE / GRID
        slat1 = SLAT1 * DEGRAD
        slat2 = SLAT2 * DEGRAD
        olon = OLON * DEGRAD
        olat = OLAT * DEGRAD

        sn = math.tan(math.pi * 0.25 + slat2 * 0.5) / math.tan(math.pi * 0.25 + slat1 * 0.5)
        sn = math.log(math.cos(slat1) / math.cos(slat2)) / math.log(sn)
        sf = math.tan(math.pi * 0.25 + slat1 * 0.5)
        sf = math.pow(sf, sn) * math.cos(slat1) / sn
        ro = math.tan(math.pi * 0.25 + olat * 0.5)
        ro = re * sf / math.pow(ro, sn)

        ra = math.tan(math.pi * 0.25 + lat * DEGRAD * 0.5)
        ra = re * sf / math.pow(ra, sn)

        theta = lon * DEGRAD - olon
        if theta > math.pi:
            theta -= 2.0 * math.pi
        if theta < -math.pi:
            theta += 2.0 * math.pi
        theta *= sn

        nx = int(ra * math.sin(theta) + XO + 0.5)
        ny = int(ro - ra * math.cos(theta) + YO + 0.5)

        return nx, ny

    def _get_base_time(self) -> tuple:
        # API 호출용 기준 시간 계산
        now = datetime.now()
        base_times = ["0200", "0500", "0800", "1100", "1400", "1700", "2000", "2300"]
        current = now.hour * 100 + now.minute

        for base_time in reversed(base_times):
            if current >= int(base_time) + 10:
                return now.strftime("%Y%m%d"), base_time

        yesterday = now - timedelta(days=1)
        return yesterday.strftime("%Y%m%d"), "2300"

    def _parse_weather_data(self, data: dict, lat: float, lon: float, nx: int, ny: int) -> dict:
        # 날씨 데이터 파싱
        items = data["response"]["body"]["items"]["item"]

        now = datetime.now()
        target_time = (now + timedelta(hours=1)).strftime("%H00")
        target_date = now.strftime("%Y%m%d")

        weather_data = {}
        for item in items:
            if item["fcstDate"] == target_date and item["fcstTime"] == target_time:
                category = item["category"]
                value = item["fcstValue"]

                if category == "TMP":
                    weather_data["temp"] = float(value)
                elif category == "REH":
                    weather_data["humidity"] = int(value)
                elif category == "SKY":
                    sky_codes = {"1": "맑음", "3": "구름많음", "4": "흐림"}
                    weather_data["sky"] = sky_codes.get(value, "알수없음")
                elif category == "PTY":
                    pty_codes = {"0": "없음", "1": "비", "2": "비/눈", "3": "눈", "4": "소나기"}
                    weather_data["precipitation"] = pty_codes.get(value, "없음")
                elif category == "POP":
                    weather_data["rain_prob"] = int(value)

        return {
            "lat": lat,
            "lon": lon,
            "grid": {"nx": nx, "ny": ny},
            "temp": weather_data.get("temp", 0),
            "humidity": weather_data.get("humidity", 0),
            "sky": weather_data.get("sky", "알수없음"),
            "precipitation": weather_data.get("precipitation", "없음"),
            "rain_probability": weather_data.get("rain_prob", 0),
            "description": f"{weather_data.get('sky', '')} ({weather_data.get('precipitation', '')})",
        }

File: frontend/src/services/weatherService.js
/*
  파일명: services/weatherService.js
  기능: 날씨 관련 서비스 (기상청 API 직접 호출)
*/

// 기상청 단기예보 API
const WEATHER_API_URL = 'https://apis.data.go.kr/1360000/VilageFcstInfoService_2.0/getVilageFcst';
const WEATHER_API_KEY = '5621048a47e5a1f37bdb05a7dd8c567dca6034fbde9af3a9cd320293cfff84dc';

/**
 * 위경도를 기상청 격자 좌표로 변환
 * Lambert Conformal Conic Projection 사용
 */
const convertToGrid = (lat, lon) => {
  const RE = 6371.00877; // 지구 반경(km)
  const GRID = 5.0; // 격자 간격(km)
  const SLAT1 = 30.0; // 표준위도1
  const SLAT2 = 60.0; // 표준위도2
  const OLON = 126.0; // 기준점 경도
  const OLAT = 38.0; // 기준점 위도
  const XO = 43; // 기준점 X좌표
  const YO = 136; // 기준점 Y좌표

  const DEGRAD = Math.PI / 180.0;

  const re = RE / GRID;
  const slat1 = SLAT1 * DEGRAD;
  const slat2 = SLAT2 * DEGRAD;
  const olon = OLON * DEGRAD;
  const olat = OLAT * DEGRAD;

  let sn = Math.tan(Math.PI * 0.25 + slat2 * 0.5) / Math.tan(Math.PI * 0.25 + slat1 * 0.5);
  sn = Math.log(Math.cos(slat1) / Math.cos(slat2)) / Math.log(sn);

  let sf = Math.tan(Math.PI * 0.25 + slat1 * 0.5);
  sf = Math.pow(sf, sn) * Math.cos(slat1) / sn;

  let ro = Math.tan(Math.PI * 0.25 + olat * 0.5);
  ro = re * sf / Math.pow(ro, sn);

  let ra = Math.tan(Math.PI * 0.25 + lat * DEGRAD * 0.5);
  ra = re * sf / Math.pow(ra, sn);

  let theta = lon * DEGRAD - olon;
  if (theta > Math.PI) theta -= 2.0 * Math.PI;
  if (theta < -Math.PI) theta += 2.0 * Math.PI;
  theta *= sn;

  const nx = Math.floor(ra * Math.sin(theta) + XO + 0.5);
  const ny = Math.floor(ro - ra * Math.cos(theta) + YO + 0.5);

  return { nx, ny };
};

/**
 * API 호출용 기준 시간 계산
 */
const getBaseTime = () => {
  const now = new Date();
  const baseTimes = ['0200', '0500', '0800', '1100', '1400', '1700', '2000', '2300'];
  const currentTime = now.getHours() * 100 + now.getMinutes();

  // 현재 시간보다 10분 이전 발표시간 찾기
  for (let i = baseTimes.length - 1; i >= 0; i--) {
    if (currentTime >= parseInt(baseTimes[i]) + 10) {
      const baseDate = now.toISOString().split('T')[0].replace(/-/g, '');
      return { baseDate, baseTime: baseTimes[i] };
    }
  }

  // 오늘 첫 발표시간 이전이면 어제 마지막 시간
  const yesterday = new Date(now);
  yesterday.setDate(yesterday.getDate() - 1);
  const baseDate = yesterday.toISOString().split('T')[0].replace(/-/g, '');
  return { baseDate, baseTime: '2300' };
};

/**
 * 날씨 데이터 파싱
 */
const parseWeatherData = (data, lat, lon, nx, ny) => {
  try {
    const items = data.response.body.items.item;

    // 현재 시간 + 1시간 (예보 데이터 조회용)
    const now = new Date();
    const targetHour = now.getHours() + 1;
    const targetTime = targetHour.toString().padStart(2, '0') + '00';
    const targetDate = now.getFullYear().toString() +
                      (now.getMonth() + 1).toString().padStart(2, '0') +
                      now.getDate().toString().padStart(2, '0');

    const weatherData = {};

    items.forEach(item => {
      if (item.fcstDate === targetDate && item.fcstTime === targetTime) {
        const category = item.category;
        const value = item.fcstValue;

        if (category === 'TMP') {
          weatherData.temp = parseFloat(value);
        } else if (category === 'REH') {
          weatherData.humidity = parseInt(value);
        } else if (category === 'SKY') {
          const skyCodes = { '1': '맑음', '3': '구름많음', '4': '흐림' };
          weatherData.sky = skyCodes[value] || '알수없음';
        } else if (category === 'PTY') {
          const ptyCodes = { '0': '없음', '1': '비', '2': '비/눈', '3': '눈', '4': '소나기' };
          weatherData.precipitation = ptyCodes[value] || '없음';
        } else if (category === 'POP') {
          weatherData.rainProb = parseInt(value);
        }
      }
    });

    return {
      latitude: lat,
      longitude: lon,
      grid_x: nx,
      grid_y: ny,
      temperature: weatherData.temp || 0,
      temp: weatherData.temp || 0,
      humidity: weatherData.humidity || 0,
      sky: weatherData.sky || '알수없음',
      precipitation: weatherData.precipitation || '없음',
      rainProbability: weatherData.rainProb || 0,
      description: `${weatherData.sky || ''} (${weatherData.precipitation || ''})`,
    };
  } catch (error) {
    console.error('[weatherService] 파싱 오류:', error);
    throw new Error('날씨 데이터 파싱 실패');
  }
};

export const weatherService = {
  /**
   * 날씨 정보 조회
   * @param {number} latitude - 위도
   * @param {number} longitude - 경도
   * @returns {Promise<Object>} 날씨 정보
   */
  getWeather: async (latitude = 37.5665, longitude = 126.9780) => {
    try {
      console.log('[weatherService] 날씨 조회 시작:', { latitude, longitude });

      // 좌표 변환
      const { nx, ny } = convertToGrid(latitude, longitude);
      console.log('[weatherService] 격자 좌표:', { nx, ny });

      // 기준 시간 계산
      const { baseDate, baseTime } = getBaseTime();
      console.log('[weatherService] 기준 시간:', { baseDate, baseTime });

      // API 요청 (serviceKey는 이미 인코딩되어 있으므로 직접 포함)
      const url = `${WEATHER_API_URL}?serviceKey=${WEATHER_API_KEY}&pageNo=1&numOfRows=100&dataType=JSON&base_date=${baseDate}&base_time=${baseTime}&nx=${nx}&ny=${ny}`;
      console.log('[weatherService] API 요청:', url);

      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'Accept': 'application/json',
        },
      });

      if (!response.ok) {
        const errorText = await response.text();
        console.error('[weatherService] API 응답 오류:', errorText);
        throw new Error(`기상청 API 오류 (${response.status})`);
      }

      const data = await response.json();
      console.log('[weatherService] API 응답:', data);

      // 응답 코드 확인
      if (data.response.header.resultCode !== '00') {
        throw new Error(`기상청 API 에러: ${data.response.header.resultMsg}`);
      }

      // 데이터 파싱
      const weatherData = parseWeatherData(data, latitude, longitude, nx, ny);
      console.log('[weatherService] 파싱된 날씨 데이터:', weatherData);

      return weatherData;
    } catch (error) {
      console.error('[weatherService] 날씨 조회 실패:', error);
      throw error;
    }
  },
};

export default weatherService;
